<a href="https://colab.research.google.com/github/chasslayy/Chasslayy-Luxe-AutoAI-project/blob/main/'Chasslayy_Luxe_AutoAI_Colab_Starter.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)

# Privacy-safe schema (NO PI/SPI)
COLUMNS = [
    "sessions_last_30d",
    "engagement_level",
    "product_category_viewed",
    "discount_used",
    "days_since_last_interaction",
    "prior_purchases",
    "avg_order_value_bucket",
    "traffic_source",
    "purchase_made"  # target
]


In [2]:
N = 800  # you can increase later

engagement_levels = ["low", "medium", "high"]
categories = ["wig", "beauty", "digital"]
sources = ["instagram", "tiktok", "pinterest", "direct", "other"]
aov_bucket = ["low", "medium", "high"]

df = pd.DataFrame({
    "sessions_last_30d": np.random.poisson(lam=6, size=N).clip(0, 40),
    "engagement_level": np.random.choice(engagement_levels, size=N, p=[0.45, 0.40, 0.15]),
    "product_category_viewed": np.random.choice(categories, size=N, p=[0.50, 0.30, 0.20]),
    "discount_used": np.random.binomial(1, 0.35, size=N),
    "days_since_last_interaction": np.random.randint(0, 61, size=N),
    "prior_purchases": np.random.poisson(lam=1.2, size=N).clip(0, 12),
    "avg_order_value_bucket": np.random.choice(aov_bucket, size=N, p=[0.40, 0.45, 0.15]),
    "traffic_source": np.random.choice(sources, size=N, p=[0.45, 0.25, 0.12, 0.10, 0.08]),
})

# Behavior-based probability of purchase (privacy-safe)
base = 0.08
p = (
    base
    + 0.03 * (df["sessions_last_30d"] / 10)
    + 0.08 * (df["prior_purchases"] / 6)
    + 0.10 * df["engagement_level"].map({"low": 0.0, "medium": 0.5, "high": 1.0})
    - 0.06 * (df["days_since_last_interaction"] / 60)
    + 0.04 * df["discount_used"]
    + 0.03 * df["traffic_source"].isin(["instagram", "tiktok"]).astype(int)
).clip(0.01, 0.95)

df["purchase_made"] = np.random.binomial(1, p)

df.head()


sessions_last_30d engagement_level product_category_viewed  discount_used  \
0                  6           medium                  beauty              1   
1                  6           medium                  beauty              0   
2                  4              low                     wig              0   
3                  5              low                  beauty              1   
4                  4              low                     wig              1   

   days_since_last_interaction  prior_purchases avg_order_value_bucket  \
0                           44                0                   high   
1                           59                0                   high   
2                           55                1                 medium   
3                           14                0                 medium   
4                            5                1                    low   

  traffic_source  purchase_made  
0      instagram              0  
1      instagram              0  
2         tiktok              0  
3         tiktok              0  
4      instagram              1

In [10]:
import numpy as np
import pandas as pd

np.random.seed(42)


In [3]:
N = 800  # you can increase later

engagement_levels = ["low", "medium", "high"]
categories = ["wig", "beauty", "digital"]
sources = ["instagram", "tiktok", "pinterest", "direct", "other"]
aov_bucket = ["low", "medium", "high"]

df = pd.DataFrame({
    "sessions_last_30d": np.random.poisson(lam=6, size=N).clip(0, 40),
    "engagement_level": np.random.choice(engagement_levels, size=N, p=[0.45, 0.40, 0.15]),
    "product_category_viewed": np.random.choice(categories, size=N, p=[0.50, 0.30, 0.20]),
    "discount_used": np.random.binomial(1, 0.35, size=N),
    "days_since_last_interaction": np.random.randint(0, 61, size=N),
    "prior_purchases": np.random.poisson(lam=1.2, size=N).clip(0, 12),
    "avg_order_value_bucket": np.random.choice(aov_bucket, size=N, p=[0.40, 0.45, 0.15]),
    "traffic_source": np.random.choice(sources, size=N, p=[0.45, 0.25, 0.12, 0.10, 0.08]),
})

# Behavior-based probability of purchase (privacy-safe)
base = 0.08
p = (
    base
    + 0.03 * (df["sessions_last_30d"] / 10)
    + 0.08 * (df["prior_purchases"] / 6)
    + 0.10 * df["engagement_level"].map({"low": 0.0, "medium": 0.5, "high": 1.0})
    - 0.06 * (df["days_since_last_interaction"] / 60)
    + 0.04 * df["discount_used"]
    + 0.03 * df["traffic_source"].isin(["instagram", "tiktok"]).astype(int)
).clip(0.01, 0.95)

df["purchase_made"] = np.random.binomial(1, p)

df.head()


sessions_last_30d engagement_level product_category_viewed  discount_used  \
0                  7             high                     wig              0   
1                  5           medium                  beauty              0   
2                  4           medium                     wig              0   
3                  5              low                     wig              0   
4                  7              low                     wig              1   

   days_since_last_interaction  prior_purchases avg_order_value_bucket  \
0                           10                0                    low   
1                           13                4                    low   
2                           46                1                 medium   
3                           24                1                   high   
4                            6                0                 medium   

  traffic_source  purchase_made  
0         tiktok              0  
1      instagram              0  
2         tiktok              0  
3          other              1  
4         direct              1

In [11]:
df = pd.DataFrame({
    "sessions_last_30d": np.random.poisson(lam=6, size=N).clip(0, 40),
    "engagement_level": np.random.choice(engagement_levels, size=N, p=[0.45, 0.40, 0.15]),
    "product_category_viewed": np.random.choice(categories, size=N, p=[0.50, 0.30, 0.20]),
    "discount_used": np.random.binomial(1, 0.35, size=N),
    "days_since_last_interaction": np.random.randint(0, 61, size=N),
    "prior_purchases": np.random.poisson(lam=1.2, size=N).clip(0, 12),
    "avg_order_value_bucket": np.random.choice(aov_bucket, size=N, p=[0.40, 0.45, 0.15]),
    "traffic_source": np.random.choice(sources, size=N, p=[0.45, 0.25, 0.12, 0.10, 0.08]),
})



In [7]:
output_path = "chasslayy_luxe_purchase_propensity.csv"
df.to_csv(output_path, index=False)
output_path


'chasslayy_luxe_purchase_propensity.csv'

In [8]:
from google.colab import files
files.download(output_path)


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

X = df.drop(columns=["purchase_made"])
y = df["purchase_made"]

cat_cols = ["engagement_level","product_category_viewed","avg_order_value_bucket","traffic_source"]
num_cols = [c for c in X.columns if c not in cat_cols]

preprocess = ColumnTransformer([
    ("cat", OneHotEncoder(handle_unknown="ignore"), cat_cols),
    ("num", "passthrough", num_cols)
])

clf = Pipeline([
    ("prep", preprocess),
    ("model", LogisticRegression(max_iter=1000))
])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
clf.fit(X_train, y_train)
pred = clf.predict(X_test)

print("Accuracy:", round(accuracy_score(y_test, pred), 4))
print("Confusion matrix:\n", confusion_matrix(y_test, pred))


Accuracy: 0.825
Confusion matrix:
 [[132   0]
 [ 28   0]]


# Task
Analyze the distribution of the 'purchase_made' target variable and re-train the Logistic Regression model using `class_weight='balanced'`. Then, evaluate the re-trained model's performance using accuracy, confusion matrix, precision, recall, and F1-score, and summarize the improvements.

## Analyze Target Variable Distribution

### Subtask:
Inspect the distribution of the 'purchase_made' target variable to confirm the class imbalance.


**Reasoning**:
To inspect the distribution of the 'purchase_made' target variable and confirm class imbalance, I need to calculate and display its value counts.



In [17]:
print(df['purchase_made'].value_counts())
print(df['purchase_made'].value_counts(normalize=True).mul(100).round(2).astype(str) + '%')

purchase_made
0    669
1    131
Name: count, dtype: int64
purchase_made
0    83.62%
1    16.38%
Name: proportion, dtype: object


**Reasoning**:
The previous code failed with a `KeyError: 'purchase_made'` because the `df` DataFrame in the kernel's memory did not contain the 'purchase_made' column, likely due to a previous cell overwriting it. To fix this, I need to regenerate the `df` DataFrame with the 'purchase_made' column before calculating its value counts.



In [13]:
N = 800

engagement_levels = ["low", "medium", "high"]
categories = ["wig", "beauty", "digital"]
sources = ["instagram", "tiktok", "pinterest", "direct", "other"]
aov_bucket = ["low", "medium", "high"]

df = pd.DataFrame({
    "sessions_last_30d": np.random.poisson(lam=6, size=N).clip(0, 40),
    "engagement_level": np.random.choice(engagement_levels, size=N, p=[0.45, 0.40, 0.15]),
    "product_category_viewed": np.random.choice(categories, size=N, p=[0.50, 0.30, 0.20]),
    "discount_used": np.random.binomial(1, 0.35, size=N),
    "days_since_last_interaction": np.random.randint(0, 61, size=N),
    "prior_purchases": np.random.poisson(lam=1.2, size=N).clip(0, 12),
    "avg_order_value_bucket": np.random.choice(aov_bucket, size=N, p=[0.40, 0.45, 0.15]),
    "traffic_source": np.random.choice(sources, size=N, p=[0.45, 0.25, 0.12, 0.10, 0.08]),
})

base = 0.08
p = (
    base
    + 0.03 * (df["sessions_last_30d"] / 10)
    + 0.08 * (df["prior_purchases"] / 6)
    + 0.10 * df["engagement_level"].map({"low": 0.0, "medium": 0.5, "high": 1.0})
    - 0.06 * (df["days_since_last_interaction"] / 60)
    + 0.04 * df["discount_used"]
    + 0.03 * df["traffic_source"].isin(["instagram", "tiktok"]).astype(int)
).clip(0.01, 0.95)

df["purchase_made"] = np.random.binomial(1, p)


print(df['purchase_made'].value_counts())
print(df['purchase_made'].value_counts(normalize=True).mul(100).round(2).astype(str) + '%')

purchase_made
0    669
1    131
Name: count, dtype: int64
purchase_made
0    83.62%
1    16.38%
Name: proportion, dtype: object


## Re-train Model with Class Weights

### Subtask:
Re-train the Logistic Regression model using `class_weight='balanced'` to inherently handle the class imbalance present in the 'purchase_made' column.


**Reasoning**:
To address the class imbalance, I will re-train the Logistic Regression model by instantiating a new model with `class_weight='balanced'` and `max_iter=1000`, creating a new pipeline with this model, fitting it to the training data, and then making predictions on the test set.



In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline

# 1. Instantiate a new LogisticRegression model with class_weight='balanced'
model_balanced = LogisticRegression(class_weight='balanced', max_iter=1000)

# 2. Create a Pipeline named clf_balanced
clf_balanced = Pipeline([
    ("prep", preprocess),
    ("model", model_balanced)
])

# 3. Fit the clf_balanced pipeline to the training data
clf_balanced.fit(X_train, y_train)

# 4. Use the clf_balanced pipeline to make predictions on the test data
pred_balanced = clf_balanced.predict(X_test)

print("Balanced model predictions generated.")

Balanced model predictions generated.


**Reasoning**:
Now that the balanced model has been trained and predictions generated, I need to evaluate its performance using various metrics: accuracy, confusion matrix, precision, recall, and F1-score. These metrics will provide a comprehensive understanding of how the model performs, especially in the context of class imbalance.



In [15]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score

# Evaluate the balanced model
accuracy_balanced = accuracy_score(y_test, pred_balanced)
conf_matrix_balanced = confusion_matrix(y_test, pred_balanced)
precision_balanced = precision_score(y_test, pred_balanced)
recall_balanced = recall_score(y_test, pred_balanced)
f1_balanced = f1_score(y_test, pred_balanced)

print("Balanced Model Evaluation:")
print(f"Accuracy: {accuracy_balanced:.4f}")
print(f"Confusion Matrix:\n{conf_matrix_balanced}")
print(f"Precision: {precision_balanced:.4f}")
print(f"Recall: {recall_balanced:.4f}")
print(f"F1-Score: {f1_balanced:.4f}")


Balanced Model Evaluation:
Accuracy: 0.5500
Confusion Matrix:
[[76 56]
 [16 12]]
Precision: 0.1765
Recall: 0.4286
F1-Score: 0.2500


**Reasoning**:
To summarize the improvements, I need to compare the performance metrics of the original model with those of the balanced model. I will first calculate and print the precision, recall, and F1-score for the original model using its predictions (`pred`) and the true labels (`y_test`).



In [16]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Evaluate the original model
accuracy_original = accuracy_score(y_test, pred)
conf_matrix_original = confusion_matrix(y_test, pred)
precision_original = precision_score(y_test, pred, zero_division=0)
recall_original = recall_score(y_test, pred, zero_division=0)
f1_original = f1_score(y_test, pred, zero_division=0)

print("\nOriginal Model Evaluation:")
print(f"Accuracy: {accuracy_original:.4f}")
print(f"Confusion Matrix:\n{conf_matrix_original}")
print(f"Precision: {precision_original:.4f}")
print(f"Recall: {recall_original:.4f}")
print(f"F1-Score: {f1_original:.4f}")


Original Model Evaluation:
Accuracy: 0.8250
Confusion Matrix:
[[132   0]
 [ 28   0]]
Precision: 0.0000
Recall: 0.0000
F1-Score: 0.0000


In [18]:
# STEP 1: Create a privacy-safe estimated order value
bucket_base = df["avg_order_value_bucket"].map({
    "low": 35,
    "medium": 75,
    "high": 140
}).astype(float)

category_boost = df["product_category_viewed"].map({
    "digital": -10,
    "beauty": 0,
    "wig": 40
}).astype(float)

noise = np.random.normal(loc=0, scale=12, size=len(df))

df["order_value_est"] = (bucket_base + category_boost + noise).clip(10, 300).round(2)

df[["avg_order_value_bucket", "product_category_viewed", "order_value_est"]].head()


avg_order_value_bucket product_category_viewed  order_value_est
0                 medium                 digital            61.84
1                 medium                  beauty            58.90
2                 medium                     wig           116.39
3                    low                     wig            82.86
4                 medium                  beauty            74.26

In [19]:
# STEP 2: Create purchase_made target
base = 0.08
p = (
    base
    + 0.03 * (df["sessions_last_30d"] / 10)
    + 0.08 * (df["prior_purchases"] / 6)
    + 0.10 * df["engagement_level"].map({"low": 0.0, "medium": 0.5, "high": 1.0})
    - 0.06 * (df["days_since_last_interaction"] / 60)
    + 0.04 * df["discount_used"]
    + 0.03 * df["traffic_source"].isin(["instagram", "tiktok"]).astype(int)
).clip(0.01, 0.95)

df["purchase_made"] = np.random.binomial(1, p)

df["purchase_made"].value_counts(normalize=True)


purchase_made
0    0.84
1    0.16
Name: proportion, dtype: float64

In [20]:
# STEP 3: Create high_value_purchase target
HIGH_VALUE_THRESHOLD = 120

df["high_value_purchase"] = (
    (df["purchase_made"] == 1) & (df["order_value_est"] >= HIGH_VALUE_THRESHOLD)
).astype(int)

df[["purchase_made", "order_value_est", "high_value_purchase"]].head(10)


purchase_made  order_value_est  high_value_purchase
0              0            61.84                    0
1              0            58.90                    0
2              0           116.39                    0
3              0            82.86                    0
4              0            74.26                    0
5              0           131.81                    0
6              0           181.74                    0
7              0            89.97                    0
8              0           128.04                    0
9              0            51.09                    0

In [21]:
df.to_csv("chasslayy_luxe_autoai_dataset.csv", index=False)


### Summary of Model Performance Comparison

**Original Model:**
- **Accuracy:** 0.8250
- **Confusion Matrix:**
  ```
  [[132   0]
   [ 28   0]]
  ```
- **Precision:** 0.0000
- **Recall:** 0.0000
- **F1-Score:** 0.0000

**Balanced Model (with `class_weight='balanced'`):**
- **Accuracy:** 0.5500
- **Confusion Matrix:**
  ```
  [[76 56]
   [16 12]]
  ```
- **Precision:** 0.1765
- **Recall:** 0.4286
- **F1-Score:** 0.2500

**Analysis of Improvements:**

1.  **Accuracy:** The accuracy of the balanced model (0.5500) is significantly lower than the original model (0.8250). However, accuracy can be misleading in imbalanced datasets as a model might achieve high accuracy by simply predicting the majority class.

2.  **Recall:** The most notable improvement is in **Recall**. The original model had a Recall of 0.0000, meaning it failed to identify any of the positive cases (`purchase_made=1`). The balanced model, on the other hand, achieved a Recall of 0.4286, indicating it can now correctly identify a substantial portion of the actual purchases. This is crucial for a business objective of identifying potential buyers.

3.  **Precision:** The original model had a Precision of 0.0000, which means it made no correct positive predictions. The balanced model improved Precision to 0.1765. While still relatively low, it signifies that the model is now making some correct positive predictions.

4.  **F1-Score:** The F1-Score, which is the harmonic mean of precision and recall, also saw a significant increase from 0.0000 in the original model to 0.2500 in the balanced model. This indicates a better balance between precision and recall, reflecting the model's improved ability to handle the positive class.

**Conclusion:** While the balanced model shows a decrease in overall accuracy, it dramatically improves the model's ability to detect the minority class (purchases). This is a desired outcome when dealing with imbalanced datasets where identifying the minority class is often the primary business objective. The original model was biased towards the majority class and essentially predicted no purchases, making it ineffective for identifying customers with high purchase propensity.

## Summary:

### Q&A
The re-trained Logistic Regression model using `class_weight='balanced'` shows significant improvements in identifying the minority class (purchases) compared to the original model. Specifically:
*   **Recall** increased from 0.0000 to 0.4286.
*   **Precision** increased from 0.0000 to 0.1765.
*   **F1-Score** increased from 0.0000 to 0.2500.

While the overall accuracy decreased from 0.8250 to 0.5500, this is a necessary trade-off to correctly identify positive cases in an imbalanced dataset, making the balanced model much more effective for its intended purpose.

### Data Analysis Key Findings
*   The 'purchase\_made' target variable exhibited significant class imbalance: 83.62% for class '0' (no purchase) and 16.38% for class '1' (purchase).
*   The original Logistic Regression model, without class balancing, performed poorly in identifying purchases, achieving an F1-Score of 0.0000, a Recall of 0.0000, and a Precision of 0.0000 for the positive class. Its confusion matrix indicated it predicted '0' for all instances.
*   The re-trained Logistic Regression model using `class_weight='balanced'` achieved an F1-Score of 0.2500, a Recall of 0.4286, and a Precision of 0.1765 for the positive class.
*   The balanced model's accuracy was 0.5500, which was lower than the original model's accuracy of 0.8250, but it demonstrated a crucial ability to identify actual purchases.

### Insights or Next Steps
*   For business objectives focused on identifying potential purchasers, the balanced model is significantly more valuable despite lower overall accuracy, as it can now detect 42.86% of actual purchases, which the original model completely missed.
*   Further optimization could involve experimenting with different `class_weight` parameters, exploring other sampling techniques (e.g., SMOTE), or trying more advanced machine learning models designed for imbalanced datasets to potentially improve precision while maintaining a good recall for the minority class.
